### PR TITLE
PDF Generator (EXPOSUREAPP-9191)

### PR DESCRIPTION
**Pull request by jurajkusnier**
~~- Implementation is not yet linked to UI, it will be possible after #4006 is merged~~
### Testing
1. generate and add some tc, vc, rc
2. open certificate detail and click "Druckversion anzeigen" from top app bar menu
~~3. check created pdf document in /data/data/de.rki.coronawarnapp.test/cache/test.pdf~~
3. click "Weiter"
4. check the PDF preview, try to share/print it and check formatting and data in the file

TODO:
- [x]  more unit tests
- [x]  wire with UI after #4006 is merged

### CODE REVIEW

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/pdf/core/Utils.kt [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+}
+
+fun Canvas.drawMultilineText(text: List<String>, paint: Paint, x: Float, y: Float) {
+    text.forEachIndexed { index, line ->
+        drawText(line, x, y + paint.textSize * index * 1.2f, paint)
+    }
+}
+
+fun ByteMatrix.toBitmap(): Bitmap = Bitmap.createBitmap(
+    IntStream.range(0, height)
```
Comment (mtwalli) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/pdf/core/Utils.kt:44]: why java streams ? can't we use Kotlin ranges?
such as 
```kotlin
    (0..height).flatMap { h ->
        (0..width).map { w ->
            if (get(w, h) > 0) Color.BLACK else Color.WHITE
        }
    }.toIntArray()
```

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/pdf/ui/poster/CertificatePosterViewModel.kt [hunk 2]
```diff
@@ -1,101 +1,75 @@
...
-                    finishPage(this)
-                }
-
-                FileOutputStream(file).use {
-                    writeTo(it)
-                    close()
+    private fun getPDFFileName(certificateData: CwaCovidCertificate): String {
+        val pdfNameSuffix =
+            with(certificateData) {
+                if (firstName != null) {
```
Comment (mtwalli) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/pdf/ui/poster/CertificatePosterViewModel.kt:46]: `firstName` could be empty too `!firstName.isNullOrEmpty` ?

Comment (mtwalli) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/pdf/ui/poster/CertificatePosterViewModel.kt:46]: BTW firstName, lastName are user language such as Russian, right? can this cause issues in file names ? should we use standradFirstName and lastName? or this by specs?

**Comment by mtwalli** [comment]
After exporting to PDF File resolution: `2480 × 3508`  , I remember we had an issue regarding PDF file resolution before for event registration , can you please double check that? https://github.com/corona-warn-app/cwa-app-android/issues/2928
and please align with iOS to see what resolution they export with? 

/cc @mlenkeit

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/pdf/ui/poster/CertificatePosterViewModel.kt (hunk 2 above)
Comment (AlexanderAlferov) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/pdf/ui/poster/CertificatePosterViewModel.kt:46]: > BTW firstName, lastName are user language such as Russian, right? can this cause issues in file names ? should we use standradFirstName and lastName? or this by specs?

Tested with:

- Лёха
- kühn
- kühn>ёжик

No issues with saving/opening. Though it is not specified in techspec I think it is really good idea to switch to standard names. Proposal (code tested):

```
val pdfNameSuffix =
            with(certificateData.rawCertificate.nameData) {
                if (givenNameStandardized != null) {
                    "_${givenNameStandardized.lowercase()}_${familyNameStandardized.lowercase()}"
                } else {
                    "_${familyNameStandardized.lowercase()}"
                }
            }
```

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/pdf/core/CertificateDrawHelper.kt [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+        val hints = mapOf(
+            /**
+             * We cannot use Charsets.UTF_8 as zxing calls toString internally
+             * and some android version return the class name and not the charset name
+             */
+            EncodeHintType.CHARACTER_SET to certificate.qrCodeToDisplay.options.characterSet.name()
+        )
+
+        val qrCode = Encoder.encode(
+            certificate.qrCodeToDisplay.content,
```
[the remark this answers is not in the document]
Comment (jurajkusnier) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/pdf/core/CertificateDrawHelper.kt:51]: fixed

**Comment by jurajkusnier** [comment]
> After exporting to PDF File resolution: `2480 × 3508` , I remember we had an issue regarding PDF file resolution before for event registration , can you please double check that? #2928
> and please align with iOS to see what resolution they export with?
> 
> /cc @mlenkeit

document size has been updated

**Review by mtwalli — APPROVED** [review]
Lgtm

**Review by AlexanderAlferov — APPROVED** [review]
🚀

**Pull request merged**